json-print/jasonkuhrt

jasonkuhrt opened the pull request.

jasonkuhrt committed:
json-print: Semi-complete solution. Missing escapes for nested quotes.

jasonkuhrt committed:
json-print: Tidy.

jasonkuhrt committed:
json-print: Setup failing sample data to work on.

jasonkuhrt committed:
json-print: Three string escape cases. Done!

jasonkuhrt committed:
json-print: doc.

jasonkuhrt committed:
json-print: Tidy.

neongreen commented on week4/json-print/jasonkuhrt/Main.hs:
```diff
@@ -0,0 +1,119 @@
+stringify (L list)   =
+    wrap "[]"
+  . List.intercalate ", "
+  . List.map stringify
```
Any specific reason why `map` is qualified but e.g. `++` isn't? I'm asking 'cause `map` is usually used unqualified, but maybe you have a good reason for it (or maybe it just looks nicer to you – it doesn't to me but that's just because I'm used to unqualified `map`).

jasonkuhrt replied:
Weird slip on my part. I usually always use `fmap`. Will fix.

neongreen commented on week4/json-print/jasonkuhrt/Main.hs:
```diff
@@ -0,0 +1,119 @@
+stringify (N number) = stringifyNumber number where
+  stringifyNumber :: (Show n, Real n) => n -> String
+  stringifyNumber n
+    | mod' n 1 == 0 = (takeWhile (/= '.') . show) n
```
You could've used `properFraction` instead.

jasonkuhrt replied:
Saves me an `import` too.

neongreen commented on week4/json-print/jasonkuhrt/Main.hs:
```diff
@@ -0,0 +1,119 @@
+
+stringify (S string) =
+    wrapQuotes
+  . escape '\\' -- Otherwise parse yields invalid/different strings. [1]
```
This doesn't escape control characters like `'\1'` as JSON spec requires. Also, I think you could've instead written something like

``` haskell
escape :: [Char] -> String -> String
escape cEscapes s =
  (\ c -> if c `elem` cEscapes then ['\\', '\\', c] else [c]) =<< s
```

to do escaping in one pass by writing `escape ['\\', '\"']`. Anyway, it won't work with control characters or newlines so it doesn't matter.

jasonkuhrt replied:
Makes sense. The revised solution changes all this anyways since as you said this was failing to escape control characters.

neongreen commented on week4/json-print/jasonkuhrt/Main.hs:
```diff
@@ -0,0 +1,119 @@
+  $ object
+  where
+  stringifyKeyValue :: String -> JSON -> String
+  stringifyKeyValue k v = wrapQuotes k ++ ":" ++ stringify v
```
This doesn't escape key values.

jasonkuhrt committed:
json-print: Fix and Refactor.

Following feedback from @neongreen.

Fix:
  * Escape keys (just like string values).
  * Escape control characters.
    Solution: JSON only allows a few control characters via
    \* syntax while all others must be represented in \u{4-hex}
    syntax. For example `\1` must be serialized into
    `\\\\u0001` but `\n` may be serialized into just `\\\\n`.

Refactor:
  * Use `fmap` over `List.map`
  * Use `properFraction` over `mod'`

jasonkuhrt commented:
@neongreen I think this is better now.

neongreen commented:
It's still not quite right.

```
> putStrLn . stringify $ S "\""
"\\""
```

The right answer would've been `"\""`.

jasonkuhrt commented:
But why not, since that's the source string? E.g.:

``` js
> JSON.stringify("\"")
'"\\""'
```

neongreen commented:
`"\""` is a string with a single character inside – a quote. Verbatim it's `"`.

In JSON, such a string looks like – verbatim – `"\""`:

```
[yom@shiny ~]$ jsc
>>> print("\"")
"
```

`putStrLn` prints a string verbatim. Therefore, when I `putStrLn` the JSON representation of a string with a single quote in it, I expect `"\""` to be printed. If I `show`-ed it instead, I'd get `"\"\\\"\""`.

`"\\""` isn't even a valid string literal:

```
[yom@shiny ~]$ jsc
>>> "\\""
Exception: SyntaxError: Unexpected EOF
```

jasonkuhrt commented:
I was about to edit with "nevermind" heh:

``` js
> console.log(JSON.stringify("\""))
"\""
```

jasonkuhrt committed:
json-print: Fix quote escape.

The issue was two-fold. First the direct problem was that
escaping a string is a matter of transmitting a literal slash prepended
to the quote, but I was supplying two.

The second issue was that I was susceptible to this problem because
my "testing" method was to take stringified values from my Haskell
demo and try to parse them in node. However I was using the value
from Haskell that I passed through `putStrLn` Which reduces the
slashes to appear correct. Ad-hoc testing methodology results in pain.

jasonkuhrt commented:
Fixed:

```
❯   putStrLn $ stringify $ S "\""
"\""
```

neongreen commented:
Same for slashes:

```
> putStrLn . stringify $ S "\\"
"\\\\"
```

(The right answer is `"\\"`.)

Also, control characters:

```
> putStrLn . stringify $ S "\1"
"\\u0001"

> putStrLn . stringify $ S "\n"
"\\n"
```

The right answers are `"\u0001"` and `"\n"`.

jasonkuhrt committed:
json-print: fix: too many parens in almost all escapes.
